docs: Add Fast Track documentation (#683)
metal3-io/metal3-docs: docs/user-guide/src/capm3/fast_track.md

# Fast Track Mode

## Overview

> Fast Track mode is an optimization feature in CAPM3 that allows BareMetalHosts

Reviewer: I would recommend mentioning here as a disclaimer that this fast track should not be confused with the Ironic fast track option and that the CAPM3 fast track and the Ironic fast track is commonly used together in Metal3 but they are different features and enabling one does not enable the other.

to skip the deprovisioning power-off cycle during machine deletion when certain
conditions are met. This can significantly speed up cluster upgrades and node
replacements by keeping hosts powered on and ready for immediate re-use.

## How It Works

When a Metal3Machine is deleted, CAPM3 normally instructs the BareMetalHost to
power off and deprovision. With Fast Track mode enabled, CAPM3 can keep the host

> powered on if automated cleaning is configured to only clean metadata.

Reviewer: Please add that the reason behind this is the following: When the automated cleaning is not disabled then the physical machine will boot into the provisioning agent to do the cleaning, in order to speed up the re-use of the physical machine fast track keeps the machine running with the agent even after the cleaning has concluded to save 1 reboot  that would happen when the machine is re-provisioned.

| false           | false            | metadata              | Off                      |
| false           | true             | metadata              | **On** (Fast Track)      |

Since PR #3106 (`Allow disable_power_off together with autoclean`, merged on
2026-03-05), `DisablePowerOff=true` can be combined with
`AutomatedCleaningMode=metadata`. In that case, `DisablePowerOff` still takes
priority and keeps the host online.

The host remains online when:

- `DisablePowerOff` is `true` (takes priority over all other settings), or
- `AutomatedCleaningMode` is set to `metadata` (not `disabled`) AND
  `CAPM3_FAST_TRACK` is set to `true`

When both conditions are met, the BareMetalHost remains powered on after the
Metal3Machine is deleted, allowing it to be quickly re-claimed by a new
Metal3Machine without waiting for a full power cycle.

## When to Use Fast Track

Fast Track is beneficial in scenarios where:

- **Rolling upgrades**: Nodes are being replaced one at a time, and speed is
  important
- **Controlled environments**: You trust that metadata cleaning is sufficient
  (no need for full disk wipes)
- **Development/testing**: Quick iteration cycles where full deprovisioning

>   adds unnecessary delay

Reviewer: In all use cases where the BMH is expected to be re-used eventually and power consumption is not a factor.

Fast Track should be avoided when:

> - **Security is paramount**: Full disk wiping is required between tenants or

Reviewer: full disk wiping is irrelevant here, BMO will be able to do full disk wiping (zeroing) as part of automated cleaning, Agent usually runs in an autonomous mode in production (or at least it should) so no one can log in, and the file system headers and partition table info is removed and usually the disk is encrypted anyhow so fast track + automated cleaning is safer than no automated cleaning at all because in that scenario the disk is fully intact and as long as the host is not re-uesd nothing will be cleaned away.

TLDR, I would not emphasize the security aspect, IMO it is not relevant here and the reasoning is also incorrect in this case.

  workloads
- **Disk cleaning is disabled intentionally**: `AutomatedCleaningMode=disabled`
  skips cleaning and therefore cannot provide metadata cleaning for Fast Track

> - **Troubleshooting**: You need hosts to fully deprovision to reset state

Reviewer: ```suggestion
- **Troubleshooting**:  physical maintenance is needed e.g. replacement of hardware that requires the machine to be powered off.
```

This reset state reasoning doesn't really mean anything to me (or in general IMO). Let's have specific examples.

## Configuration

### Enabling Fast Track

Fast Track is configured via the `CAPM3_FAST_TRACK` environment variable on the
CAPM3 controller. It defaults to `false`.

#### Via clusterctl

Add to your clusterctl configuration file:

```yaml
variables:
  CAPM3_FAST_TRACK: "true"
```

#### Via Environment Variable

If deploying the controller directly, set the environment variable:

```bash
export CAPM3_FAST_TRACK=true
```

### Configuring AutomatedCleaningMode

For Fast Track to work, your BareMetalHost resources must have

> `AutomatedCleaningMode` set to `metadata`. In CAPM3 workflows, configure this

Reviewer: Will most likely change in the future, but anything other than disabled should work, just FYI. Not 100% sure it should be mentioned but it could be.

through the Metal3MachineTemplate or Metal3Machine fields described in the
[CAPM3 automated cleaning documentation](./automated_cleaning.md), and let CAPM3
replicate the setting to the BareMetalHost. This can also be combined with
`DisablePowerOff=true`; in that case `DisablePowerOff` remains the deciding
factor for keeping the host online.

## Behavior During Machine Deletion

When a Metal3Machine is being deleted:

1. CAPM3 clears the BareMetalHost's image, customDeploy, userData, metaData,
   and networkData references
1. Based on the AutomatedCleaningMode and CAPM3_FAST_TRACK values, CAPM3 sets
   the BMH's `Online` field:
   - **Fast Track active**: Host stays online (`Online: true`)
   - **Fast Track inactive**: Host is powered off (`Online: false`)
1. CAPM3 waits for the host to reach an available state before fully releasing
   it
1. The ConsumerRef and other association data are cleared

## Monitoring

> When Fast Track is active, you'll see log messages like:

Reviewer: ```suggestion
When Fast Track is active, see log messages will be produced by capm3 like:
```

```text
Set host Online field based on DisablePowerOff, AutomatedCleaningMode, and Capm3FastTrack host=node-0 automatedCleaningMode=metadata hostSpecOnline=true
```

When Fast Track keeps a host online, the BareMetalHost will transition directly

> from `Provisioned` to `Available` without going through `Deprovisioning` with a

Reviewer: I feels this sentence is a bit confusing, reader could get an impression that the deprovisioning state is skipped even though deprovisioning will still happen. Please try to phrase this a bit more clearly.

# Fast Track Mode

## Overview

Fast Track mode is an optimization feature in CAPM3 that keeps BareMetalHosts
powered on during machine deletion when certain conditions are met. This can
significantly speed up cluster upgrades and node replacements by avoiding an
extra power cycle before the host is re-used.

CAPM3 Fast Track is different from Ironic Fast Track. Ironic Fast Track reuses
an already-running agent ramdisk across Ironic inspection, cleaning, and
deployment. CAPM3 Fast Track controls whether CAPM3 leaves the BareMetalHost
`Online` during Metal3Machine deletion. These features are commonly used
together in Metal3, but enabling one does not enable the other.

## How It Works

When a Metal3Machine is deleted, CAPM3 normally lets the BareMetalHost power off
as part of deprovisioning. With Fast Track enabled, CAPM3 can keep the host

> powered on when automated cleaning is not disabled.

Reviewer 2: nit: double negation can be hard to understand. This could be simplified as

```suggestion
powered on when automated cleaning is enabled.
```

On further thought I guess this wording was chosen based on the actual field value (`disabled`)? Maybe we could clarify it as

```suggestion
powered on when automated cleaning is set to anything other than `disabled`.
```

re-provisioned.

The behavior is controlled by three factors:

1. **BareMetalHost.Spec.DisablePowerOff**: If `true`, the host always stays
   online (highest priority)
1. **CAPM3_FAST_TRACK environment variable**: Set to `true` or `false`
1. **BareMetalHost.Spec.AutomatedCleaningMode**: Set to `disabled` or a
   cleaning-enabled mode such as `metadata`

### Behavior Matrix

| DisablePowerOff | CAPM3_FAST_TRACK | AutomatedCleaningMode         | BMH Online Status        |
| --------------- | ---------------- | ----------------------------- | ------------------------ |
| true            | any              | any                           | **On** (DisablePowerOff) |
| false           | false            | disabled                      | Off                      |
| false           | true             | disabled                      | Off                      |
| false           | false            | not disabled (metadata today) | Off                      |
| false           | true             | not disabled (metadata today) | **On** (Fast Track)      |

Since PR #3106 (`Allow disable_power_off together with autoclean`, merged on

> 2026-03-05), `DisablePowerOff=true` can be combined with

Reviewer 2: We should convert this to release tags. In what release is/will this be included? Users should not need to know about individual PRs.
